rust/otap-dataflow/crates/quiver/src/wal/writer.rs

```diff
@@ -0,0 +1,1174 @@
+    rotated_files: VecDeque<RotatedWalFile>,
+    /// Next rotation id to use when rotating (initialized to max existing + 1).
+    next_rotation_id: u64,
+    /// Most recent offset validated to be on an entry boundary in the active file.
+    active_file_checkpoint_offset: u64,
+    /// Sequence number associated with the last committed consumer checkpoint.
+    last_checkpoint_sequence: Option<u64>,
+}
+
+/// Opaque marker returned after an append.
+///
+/// Contains the byte position and sequence number of the written entry.
+/// Useful for correlating writer positions with reader cursors during testing
+/// or debugging. Operators typically don't need to inspect these values.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(crate) struct WalOffset {
+    /// Byte offset where the entry starts in the active WAL file.
+    pub position: u64,
+    /// Monotonically increasing sequence number assigned to this entry.
+    pub sequence: u64,
+}
+
+impl WalWriter {
+    pub fn open(options: WalWriterOptions) -> WalResult<Self> {
+        if let Some(parent) = options.path.parent() {
+            std::fs::create_dir_all(parent)?;
+        }
+
+        let mut file = OpenOptions::new()
+            .create(true)
+            .read(true)
+            .write(true)
+            .truncate(false)
+            .open(&options.path)?;
+
+        let metadata = file.metadata()?;
+        let metadata_len = if metadata.len() == 0 {
+            let header = WalHeader::new(options.segment_cfg_hash);
+            header.write_to(&mut file)?;
+            file.flush()?;
+            WAL_HEADER_LEN as u64
+        } else if metadata.len() < WAL_HEADER_LEN as u64 {
+            return Err(WalError::InvalidHeader("file smaller than header"));
+        } else {
+            let header = WalHeader::read_from(&mut file)?;
+            if header.segment_cfg_hash != options.segment_cfg_hash {
+                return Err(WalError::SegmentConfigMismatch {
+                    expected: options.segment_cfg_hash,
+                    found: header.segment_cfg_hash,
+                });
+            }
+            metadata.len()
+        };
+
+        let _ = file.seek(SeekFrom::End(0))?;
+        let sidecar_path = checkpoint_sidecar_path(&options.path);
+        let checkpoint_state = load_checkpoint_state(&sidecar_path)?;
+
+        let active_file = ActiveWalFile::new(file, metadata_len);
+        let mut coordinator =
+            WalCoordinator::new(options, sidecar_path, checkpoint_state, metadata_len);
+        coordinator.reload_rotated_files(active_file.len())?;
+        coordinator.restore_checkpoint_offsets(active_file.len());
+
+        let mut writer = Self {
+            active_file,
+            coordinator,
+            next_sequence: 0,
+            #[cfg(test)]
+            test_crashed: false,
+        };
+        writer.next_sequence = writer.coordinator.detect_next_sequence()?;
+        Ok(writer)
```

Comment:
Would this be an issue if collector/process crashed mid-write, the WAL file will have a partial/corrupt record at the end. After restart, `WalWriter::open()` will seek to physical EOF, and subsequent writes happening after corrupted data could result in:
```
Before crash:  [Batch 1] [Batch 2] [Batch 3] [partial...]
After restart: [Batch 1] [Batch 2] [Batch 3] [garbage] [Batch 4] [Batch 5]
```

Either way, would be good to have a test that simulates a partial write and verifies recovery behavior (whether that's truncation, detection, or something else).